design-proposals/instancetype.kubevirt.io/expand-during-submission.md

```diff
@@ -0,0 +1,301 @@
+  instance types and preferences
+
+## Non Goals
+
+* The default behavior will not change as part of this work
+
+## User Stories
+
+* As a cluster owner I want to configure the expansion of all new
+  `VirtualMachines` referencing instance types and preferences
+* As a VirtualMachine owner I want to optionally enable or disable the expansion
+  my VirtualMachine referencing an instance type or preference
+
+## Repos
+
+* kubevirt/kubevirt
+
+## Design
+
+A new `KubeVirt` configurable will be introduced to have `VirtualMachines`
+referencing instance types or preferences expanded.
+
+Additional `VirtualMachine` configurables will also be introduced to allow a
+specific newly submitted VirtualMachine referencing instance types or
+preferences to be expanded.
+
+Both sets of configurables will default to `reference`.
+
+The `VirtualMachine` configurable will override the cluster-wide `KubeVirt`
+configurable in situations where they are not equal.
+
+When requested the `VirtualMachine` controller will expand any referenced
+instance type or preference when the `VirtualMachine` is reconciled.
+
+Note that this will only occur when no revision has already been collected for
+the resource and as such wouldn't impact existing `VirtualMachines` unless a
+user explicitly requests to expand by removing the existing reference to the
+revision.
+
+## Concerns
+
+### Exposing users to API complexity within VirtualMachines
+
+One of the original design goals with instance types and preferences was to
+simplify creation by reducing a users exposure to the core
+`VirtualMachineInstanceSpec` API and thus their overall decision matrix.
+
+While this proposal doesn't change the ability to simplify creation of
+`VirtualMachines` it can result in a fully flattened `VirtualMachine` exposing
+all of this complexity at that level once again.
+
+### Breaking declarative management of VirtualMachines using instance types
+
+This expansion behavior will break any declarative management of these
+`VirtualMachines` as they will substantially change after initial submission.
+Users will need to explicitly request to not expand their `VirtualMachines`
+referencing instance types or preferences to avoid this.
+
+## Alternatives
+
+### Eventual expansion of *all* VMs
+
+A possible alternative to this design proposal is to have the VirtualMachine
+controller eventually expand all VirtualMachines referencing instance types or
+preferences regardless of a revision already being stored.
+
+This behaviour could be controlled by the same cluster-wide and VirtualMachine
+specific configurables as suggested by this proposal.
+
+This could however result in drastic changes to existing VirtualMachine objects
+without user interaction.
+
+We would also still need to retain webhook validation logic to ensure that a
+given `VirtualMachine` using an instance type or preference is valid during
+submission.
+
+This would also put additional load on the controller if the cluster-wide
```

Comment:
```suggestion
This would also put one-off load on the controller if the cluster-wide
```
I think that on-startup load is reduced by approximately same amount. This option just makes it happen sooner.

Reply:
ACK yes fair point, changed.